Improve image push workflows

piyush-garg opened the pull request:
This will now do an image build and push, only if push
happened on main branch and nightly branch. Previously this
was doing on all the branches which is not required and
release specific branches is handled in the other
image push workflow with multi arch

This will improve the multi arch image push workflow, previously it was
pushing image to latest and tag only, and we were ending up pushing no
image for version like v0.15.x and also images were not containing
the last version change commit

Now the release images will be pushed by multiarch image push workflow,
it will be now triggered on push on release branches instead of tag
so that it contains the last version change commit, also these will
run parallel for example like v0.15.0, v0.15.x, stable and latest
so that images will be available soon after release. Also latest will be
built and pushed when changes to stable will be pushed so we dont end up
pushing to latest for old bug fix release that was the case before this
refactoring

# Submitter Checklist

- [ ] 📝 Please ensure your commit message is clear and informative. For guidance on crafting effective commit messages, refer to the How to write a git commit message guide. We prefer the commit message to be included in the PR body itself rather than a link to an external website (ie: Jira ticket).

- [ ] ♽ Before submitting a PR, run make test lint to avoid unnecessary CI processing. For an even more efficient workflow, consider installing [pre-commit](https://pre-commit.com/) and running pre-commit install in the root of this repository.

- [ ] ✨ We use linters to maintain clean and consistent code. Please ensure you've run make lint before submitting a PR. Some linters offer a --fix mode, which can be executed with the command make fix-linters (ensure [markdownlint](https://github.com/DavidAnson/markdownlint) and [golangci-lint](https://github.com/golangci/golangci-lint) tools are installed first).

- [ ] 📖 If you're introducing a user-facing feature or changing existing behavior, please ensure it's properly documented.

- [ ] 🧪 While 100% coverage isn't a requirement, we encourage unit tests for any code changes where possible.

- [ ] 🎁 If feasible, please check if an end-to-end test can be added. See [README](https://github.com/openshift-pipelines/pipelines-as-code/blob/main/test/README.md) for more details.

- [ ] 🔎 If there's any flakiness in the CI tests, don't necessarily ignore it. It's better to address the issue before merging, or provide a valid reason to bypass it if fixing isn't possible (e.g., token rate limitations).

piyush-garg commented:
But now i am thinking, as we push the version change in branch, instead of tag, i should make the release wrkflow to run on release branches and stable. And grab the version from there.

piyush-garg commented:
> But now i am thinking, as we push the version change in branch, instead of tag, i should make the release wrkflow to run on release branches and stable. And grab the version from there.

I need to check if it can be done in metadata action function, and we may need to create a separate workflow for `stable`

chmouel commented:
do you want us to merge this @piyush-garg

piyush-garg commented:
> do you want us to merge this @piyush-garg

This will fix the current bug that we are not pushing to `v0.25.x'

Now the image build will happen on `main` and `nightly` if there is any push previously that was happening for all branches. Creating unnecessary images like `release-v0.25.x` / `release-v0.25.0`

Also the multi arch images will be done after pushing a tag, and push 4 images - `latest`, `v0.25.x`, `v0.25.0` and `stable`

The only thing is  - the images pushed will not have commit like this one https://github.com/openshift-pipelines/pipelines-as-code/commit/edcfa0284ba21dc3a0ebdf8e754810156976999b which is the case right now also.

piyush-garg commented:
I will rework on this today to get this in final shape.

piyush-garg commented:
This is good to review now @chmouel @savitaashture

chmouel commented:
I am not keen to have two different workflow that is duplicating most of the things, 

can't you use conditionals ?

piyush-garg commented:
> I am not keen to have two different workflow that is duplicating most of the things,
> 
> can't you use conditionals ?

checking on this

piyush-garg committed:
Improve ghcr image push workflow

This will now do an image build and push, ony if push
happened on main branch and nightly branch. Previously this
was doing on all the branches which is not required and
release specific branches is handled in the other
image push workflow with multi arch

piyush-garg committed:
Improve multiarch image push workflow on release

This will improve the multi arch image push workflow, previously it was
pushing image to latest and tag only, and we were ending up pushing no
image for version like v0.15.x and also images were not containing
the last version change commit

Now the release images will be pushed by multiarch image push workflow,
it will be now triggered on push on release branches instead of tag
so that it contains the last version change commit, also these will
run parallel for example like v0.15.0, v0.15.x, stable and latest
so that images will be available soon after release. Also latest will be
built and pushed when changes to stable will be pushed so we dont end up
pushing to latest for old bug fix release that was the case before this
refactoring

piyush-garg commented:
> I am not keen to have two different workflow that is duplicating most of the things,
> 
> can't you use conditionals ?

Updated the PR by merging the workflows into one. Good to review now. Thanks

chmouel commented:
sounds good to me, but the only to test this is by doing a release right?

(and good job to keep the change minimal)

piyush-garg commented:
> sounds good to me, but the only to test this is by doing a release right?
> 
> (and good job to keep the change minimal)

I have tested this on my fork, with pushing release, stable, main branches etc Everything worked fine there, we can watch it for next release also

chmouel commented:
k cool thanks
